File: leetcode/editor/cn/191_位1的个数_test.java
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.BDDAssertions.then;


class NumberOf1BitsTest {

    final Solution solution = new Solution();

    @Test
    void givenNormal_when_thenSuccess() {
        then(solution.hammingWeight(3)).isEqualTo(2);
        then(solution.hammingWeight(-1)).isEqualTo(32);
    }

    //leetcode submit region begin(Prohibit modification and deletion)
    public class Solution {

        // you need to treat n as an unsigned value
        public int hammingWeight(int n) {
            int ans = 0;
            while (n != 0) {
                n &= (n - 1);
                ans += 1;
            }
            return ans;
        }

    }
//leetcode submit region end(Prohibit modification and deletion)


}

File: leetcode/editor/cn/191_位1的个数_test.py
import unittest


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        self.assertEqual(self.solution.hammingWeight(3), 2)
        self.assertEqual(self.solution.hammingWeight(-1), 32)


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def hammingWeight(self, n: int) -> int:
        # 无法统计负数
        # return f'{n:b}'.count('1')
        ans: int = 0
        for x in range(32):
            if n == 0:
                break
            n &= n - 1
            ans += 1
        return ans


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
